### PR TITLE
Implement a center formatting sequence

### PR DESCRIPTION
**Pull request by dj-bauer**
Seems like I got it working and wrote a test and documentation for it.

### CODE REVIEW

**Comment by Minoru** [comment]
:wave: Looks like you're applying the formatting manually. If you have `rustfmt` installed, you can just run `cargo fmt`. `rustfmt` can be installed via `rustup` (and *is* installed by default for the last ten or twenty Rust releases).

**Comment by dj-bauer** [comment]
>  Looks like you're applying the formatting manually. If you have `rustfmt` installed, you can just run `cargo fmt`. `rustfmt` can be installed via `rustup` (and _is_ installed by default for the last ten or twenty Rust releases).

Thanks I just read it when I saw, the format check passed. :D

**Comment by Minoru** [comment]
Thanks for the PR! Before getting to the code review, let's discuss how this is supposed to work. I have two questions:

1. what's the purpose of "width" in `%=[char][width]`?

    I thought it would enable me to create something like a table, with columns of certain width, using a sequence like this: `%==11first%=x16SECOND%=~13third`. I expected `===first===xxxxxSECONDxxxxx~~~~third~~~~`, but tests returned `firstSECOND~~~~third~~~~`. Am I misunderstanding this feature?

2. the code seem to support negative widths (uses `isize`, accepts minus), but the doc doesn't mention this. What's the idea behind that?

Just so you're aware: the next Newsboat release is scheduled for the next weekend, and since this feature isn't a blocker, I won't hold off a release for it.

**Comment by dj-bauer** [comment]
Oh, it seems like I have to take a look the code again. The way you expected it to work is how it is supposed to work. My plan was to use a width of 0 for the whole terminal width. Less than 0 is not intentional. I am sorry for all this probably stupid things, but I have never written code for/with someone  else than me, so I am currently learning a lot.

**Comment by Minoru** [comment]
@dj-bauer, just so you're aware: GitHub won't notify me about a force-push, so if you want me to take a look, simply ping me

 Also don't hesitate to ping if there's something I can help you with. I know out strwidth functions can be a bit confusing, for example. You don't have to figure out the whole of Newsboat just to submit a PR ;)

**Comment by dj-bauer** [comment]
@Minoru I did not really make any progress regarding the first question you mentioned.

> 1. what's the purpose of "width" in `%=[char][width]`?
>    I thought it would enable me to create something like a table, with columns of certain width, using a sequence like this: `%==11first%=x16SECOND%=~13third`. I expected `===first===xxxxxSECONDxxxxx~~~~third~~~~`, but tests returned `firstSECOND~~~~third~~~~`. 

I think it has to do with the way how the arguments for sequences are parsed. I kinda copied the parsing from padded_format() , but I think this is where the issue comes from. I dont really know though, what I am missing for that.

> 2. the code seem to support negative widths (uses `isize`, accepts minus), but the doc doesn't mention this. What's the idea behind that?

I think I have fixed that in my latest commit.

**Comment by Minoru** [comment]
Actually, now that I think more about this, I realize that my example above was wrong. If I wanted a string of 11 characters with word "first" in between equals signs, I'd just write `===first===` by hand—no new format specifiers needed for that. So let's forget about that example, and start our analysis afresh.

> I think it has to do with the way how the arguments for sequences are parsed. I kinda copied the parsing from padded_format() , but I think this is where the issue comes from.

Oh, that's right! I think you correctly identify the problem with the current design: the format specifier doesn't "know" what text it should format. Our existing specifiers always know that: `%[char]` requires a character, `%?` uses `&` and `?` to delimit "branches", and `%>` simply applies to everything that comes after it. The new `%=` specifier currently applies to everything after it, too, but that's not what we want.

As I see it, specifiers can be split into two categories:
1. *values* that get replaces by some text. `%[char]` and `%?` are in this category;
2. *layouts* that get replaced by a configurable "filler character". `%>` is in this category.

So now we have to decide which category `%=` belongs to (or create a new one).

Should it be a value? It makes sense: `%-[width][char]` provides left padding for `[char]`, `%[width][char]` provides right padding, so we can add `%=[width][char]` to provide centring. (Note that `[char]` here is an identifier, like `T` for title or something).

Or should `%=` be a layout? In that case, it needs delimiters to demarcate where the "content" starts and ends. Parentheses, for example: `%=21x(%t and %T and %F too)` means "format `%t and %T and %F too` as you would usually, then centre them inside a box of 21 characters, using `x` character for padding".

A layout is more generic, but I'm not convinced that the added complexity is worth that. What are your thoughts, @dj-bauer? Is the first option (value) sufficient for your needs? Does it sound like something you can implement?

**Comment by dj-bauer** [comment]
> Actually, now that I think more about this, I realize that my example above was wrong. If I wanted a string of 11 characters with word "first" in between equals signs, I'd just write `===first===` by hand—no new format specifiers needed for that. So let's forget about that example, and start our analysis afresh.

Ah, that is why I didn't notice it in my tests using format specifiers.

> A layout is more generic, but I'm not convinced that the added complexity is worth that. What are your thoughts, @dj-bauer? Is the first option (value) sufficient for your needs? Does it sound like something you can implement?

I think the first option suits the case well, because as you mentioned when you don't use an format specifier you  can simply write it out. That is the for me working implementation. The only thing actually left to do would be formatting I think then.

**Comment by Minoru** [comment]
I'm not sure I understood you there. I take it that:

1. you support the first option (value);
2. the code in this PR does what you want.

But the code here doesn't implement the first option, since it doesn't take an identifier like `%[char]` does. So I'm confused.

As far as I understand, to declare the first option implemented we need code that passes a test like this:

```rust
let mut fmt = FmtStrFormatter::new();
fmt.register_fmt('T', "whatever");
assert_eq!(fmt.do_format("%=20T", 0), "      whatever      ");
```

**Comment by Minoru** [comment]
@dj-bauer, ping? Has this PR stalled, or are you just distracted with some real-life stuff?

**Comment by dj-bauer** [comment]
Oh, yeah I was kinda distracted privately, but I will finish this in the next one or two weeks

**Comment by dj-bauer** [comment]
Initially I had the idea to specify the filling character instead of just filling with whitespace.
Therefore it would have to pass a test like
`let mut fmt = FmtStrFormatter::new();
fmt.register_fmt('T', "whatever");
assert_eq!(fmt.do_format("%= 20T", 0), "      whatever      ");
assert_eq!(fmt.do_format("%==20T", 0), "======whatever======");
assert_eq!(fmt.do_format("%=-20T", 0), "------whatever------");``
Or would that contradict the philosophy of these formatting sequences?

**Comment by dj-bauer** [comment]
I think I got everything good except for the part with formatting the actual identifier. I mostly copied the code from format_spacing, but I cannot figure out, why I don't get the identifier e.g. t formatted.
here is an example for what I get:
```rust
let mut fmt = FmtStrFormatter::new();
fmt.register_fmt('T', "whatever");
assert_eq!(fmt.do_format("%= 20T", 0), "         T        ");
assert_eq!(fmt.do_format("%==20T", 0), "=========T========");
assert_eq!(fmt.do_format("%=-20T", 0), "---------T--------");
```

**Comment by Minoru** [comment]
> Initially I had the idea to specify the filling character instead of just filling with whitespace. […] Or would that contradict the philosophy of these formatting sequences?

Yes, I think it would get out of line with the existing sequences. To be clear, I'm fine with adding filling characters to the "value" formats, it's just that:
1) the new format should be backwards-compatible with the current one;
2) this should done for *all* "value" formats, not just this new one.

So what direction do you want to take with this PR?

> I mostly copied the code from format_spacing, but I cannot figure out, why I don't get the identifier e.g. t formatted.

`format_spacing` is a "layout", it doesn't process identifiers. It's better to take inspiration from `format_format`, which processes things like `%-20T`. Actually, `format_format` is the best place to add the new code: add a new `Center` constant to `Padding` enum, tweak the parser to emit a `Specifier::Format` for `%=20T`, and add a new branch to the `match` inside `format_format()`.

BTW, I'll be busy from tomorrow until Monday (both inclusive), so please don't take it amiss if I don't review anything or respond to comments during that period. I'll get back to everything on Tuesday.

**Comment by Minoru** [comment]
Forgot to say one important thing: thanks for sticking with this PR! Especially considering it's your first one to this project. I'm sorry I can't just merge this as-is, but it's a user-visible feature which will be impossible to take out other than via major version bump. I don't plan to release 3.0 just yet, so I'd like to get the design as polished as I can. :heart:

**Comment by dj-bauer** [comment]
> So what direction do you want to take with this PR?

I would skip the idea with setting the filling character then propably.

> `format_spacing` is a "layout", it doesn't process identifiers. It's better to take inspiration from `format_format`, which processes things like `%-20T`. Actually, `format_format` is the best place to add the new code: add a new `Center` constant to `Padding` enum, tweak the parser to emit a `Specifier::Format` for `%=20T`, and add a new branch to the `match` inside `format_format()`.

Thank you for the idea. I just implemented that and it is much nicer.
Now it passes the test you mentioned earlier like that
```rust
let mut fmt = FmtStrFormatter::new();
fmt.register_fmt('T', "whatever");
assert_eq!(fmt.do_format("%=20T", 0), "      whatever      ");
```

File: rust/libnewsboat/src/fmtstrformatter/parser.rs [hunk 1]
```diff
@@ -13,6 +13,7 @@ pub enum Padding {
     Left(usize),
     /// Pad the value on the right until it reaches specified width.
     Right(usize),
+    Center(usize),
```
Comment (Minoru) [inline rust/libnewsboat/src/fmtstrformatter/parser.rs:16]: Please add a documentation comment here.

File: doc/newsboat.asciidoc [hunk 2]
```diff
@@ -1038,6 +1038,7 @@ sequence identifier `[char]` contains any non-whitespace characters, then
 `[format 1]` will be evaluated and inserted, otherwise `[format 2]` will be
 evaluated and inserted. The `&` and `[format 2]` are optional, i.e. if the
 identifier's text is empty, then an empty string will be inserted.
+Newsboat also provides a way center a string in a given width and with any given whitespace character with `%=[char][width]`. If width it = 0 the whole width of the window will be used.
```
Comment (Minoru) [inline doc/newsboat.asciidoc:1041]: We changed the format, so please update the documentation here to match.

File: rust/libnewsboat/src/fmtstrformatter/mod.rs [hunk 3]
```diff
@@ -119,6 +119,26 @@ impl FmtStrFormatter {
                 result.push_str(text);
                 result.push_str(&padding);
             }
+
+            Padding::Center(total_width) => {
+                let mut w = total_width;
+                if total_width == 0 {
+                    w = _width as usize;
```
Comment (Minoru) [inline rust/libnewsboat/src/fmtstrformatter/mod.rs:126]: Underscore means that the variable is not used. Since it *is* used now, please rename it to `width`. (I expect this is something Clippy would warn you about once you've rebased onto current `master` branch).

File: rust/libnewsboat/src/fmtstrformatter/mod.rs [hunk 4]
```diff
@@ -190,6 +210,14 @@ impl FmtStrFormatter {
 mod tests {
     use super::*;
 
+    #[test]
+    fn t_do_format_center() {
+        let mut fmt = FmtStrFormatter::new();
+        fmt.register_fmt('T', "whatever".to_string());
+        assert_eq!(fmt.do_format("%=20T", 0), "      whatever      ");
+        assert_eq!(fmt.do_format("%=19T", 0), "     whatever      ");
```
Comment (Minoru) [inline rust/libnewsboat/src/fmtstrformatter/mod.rs:218]: It'd be nice to have tests for cases where padding width is less than the value's length, e.g. `%=3T` where `T` is `"whatever"`. Don't forget to add the same to the C++ test.

File: rust/libnewsboat/src/fmtstrformatter/parser.rs [hunk 5]
```diff
@@ -45,6 +46,17 @@ fn spacing(input: &str) -> IResult<&str, Specifier> {
     Ok((input, Specifier::Spacing(chr)))
 }
 
+fn center_format(input: &str) -> IResult<&str, Specifier> {
+    let (input, _) = tag("%=")(input)?;
+    let (input, width) = take_while(|chr: char| chr.is_ascii() && (chr.is_numeric()))(input)?;
+    let (input, format) = take(1usize)(input)?;
+
+    let format = format.chars().next().unwrap();
```
Comment (Minoru) [inline rust/libnewsboat/src/fmtstrformatter/parser.rs:54]: Please add a comment explaining why this `unwrap()` would never panic. (Panic crashes the program, so we try to avoid those, but sometimes we are reasonably sure a panic can't happen — we write a comment to explain our reasoning, then)

Comment (dj-bauer) [reply rust/libnewsboat/src/fmtstrformatter/parser.rs:56]: I added the comment and thanks for the detailed explanation why the comment is necessary/convention :)

File: rust/libnewsboat/src/fmtstrformatter/mod.rs (hunk 4 above)
Comment (dj-bauer) [reply rust/libnewsboat/src/fmtstrformatter/mod.rs:218]: I added the test and actually fixed it to pass the test as well

File: rust/libnewsboat/src/fmtstrformatter/parser.rs (hunk 1 above)
Comment (dj-bauer) [reply rust/libnewsboat/src/fmtstrformatter/parser.rs:17]: Done. Hopefully descriptive enough

File: doc/newsboat.asciidoc (hunk 2 above)
Comment (dj-bauer) [reply doc/newsboat.asciidoc]: Is updated and broken down into two smaller sentences instead of one huge one.

File: doc/newsboat.asciidoc [hunk 6]
```diff
@@ -1038,7 +1038,7 @@ sequence identifier `[char]` contains any non-whitespace characters, then
 `[format 1]` will be evaluated and inserted, otherwise `[format 2]` will be
 evaluated and inserted. The `&` and `[format 2]` are optional, i.e. if the
 identifier's text is empty, then an empty string will be inserted.
-Newsboat also provides a way center a string in a given width and with any given whitespace character with `%=[char][width]`. If width it = 0 the whole width of the window will be used.
+Newsboat also provides a way to center a string in a given width followed by the actual sequence identifier, which is usually a single letter. It is done through `%=[width][identifier]`. If width it = 0 the whole width of the window will be used.
```
Comment (Minoru) [inline doc/newsboat.asciidoc:1041]: It actually makes sense to move that to the previous paragraph. You can then avoid mentioning that a sequence identifier is a single letter: just show the format, say it centers the value, and mention the "width=0" corner case.

By the way: sorry I thought of that so late, but can you please add a test for that corner case?

**Comment by Minoru** [comment]
Thanks! #1100 takes longer than anticipated, sorry about that. Please make the CI here pass, and I can do the rebase myself once #1100 is merged.

File: doc/newsboat.asciidoc (hunk 6 above)
Comment (dj-bauer) [reply doc/newsboat.asciidoc:1041]: I don't exactly how to manage that test, when the width is then dynamically defined by the window with and I therefore cannot compare to a reference string.

Comment (Minoru) [reply doc/newsboat.asciidoc:1041]: Tests call `do_format()`, the second argument of which is the window width. That value travels through the calls and ends up in the `width` in `format_format()`. So just pass something other than zero as a second argument in a test.

Comment (dj-bauer) [reply doc/newsboat.asciidoc]: I added the corresponding test and fixed the documentation.

**Comment by Minoru** [comment]
Merged! Thanks for the pull request, @dj-bauer!
